### ISSUE TITLE
Table of contents block – links don't work on initial page save

### ISSUE THREAD

**Issue by Greatdane** [WordPress/gutenberg#43595]
### Description

If you add the Table of Contents block to a new page, when you publish the page, the links don't work properly and instead open a new page with a Page not found error. The URL is retaining an **auto-draft** in it. If you edit and Update the page, the links will then work properly. Alternatively, if you publish the page first, then add the Table of Contents block, they will work correctly.

### Step-by-step reproduction instructions

Create a new page
Add content, with headings, and the Table of Content block
Publish the page
View the page
Try the TOC links
Return to the editor and make a change
Update the published page
View the page
The TOC links now work

### Screenshots, screen recording, code snippet

https://user-images.githubusercontent.com/42655541/185758215-2ba7418d-aed8-4e33-a008-327fb0fa8ae8.mp4

### Environment info

WordPress version 6.0.1
Gutenberg 13.9.0

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by aatanasovdev** [comment]
I was able to reproduce this issue following the steps. 

The tricky part is that the public URL is generated after the publish action.

When a page is added, the links of the headings are set correctly to the block's store without refreshing the edit page, but the page requires to be updated to use the public URL inside its content.

As a solution, we can remove the page URL from the links and keep only the generated hashtags.

For example, from `https://example.com/sample-page/#heading-title` to `#heading-title`. 

I'm unable to find information on whether there was a reason for using the full URLs. Such page links work with hashtags only, and this solution might be sufficient.

**Comment by aaronrobertshaw** [comment]
Following the replication steps outlined in this issue, the ToC links worked for me on initial save. @Greatdane Could you confirm if this issue has been resolved or expand on the test steps?


https://user-images.githubusercontent.com/60436221/187351907-ef79f49f-3655-4bdc-98c1-c8eab16ea5e0.mp4

**Comment by aatanasovdev** [comment]
@aaronrobertshaw, which permalink structure did you use while recording the video? 

If the **Plain** permalink structure is used, the block works. 

However, it doesn't work for me when using other permalink structures such as **Post Name**. In that case, the links don't include the page's name, but `auto-draft` is added instead.

<img width="177" alt="Screenshot 2022-08-30 at 8 22 44" src="https://user-images.githubusercontent.com/38669058/187356059-eab935e5-99ae-40aa-b748-f00772172e7e.png">

**Comment by aaronrobertshaw** [comment]
Hi @aatanasovdev, I tested with both **Plain** and **Post Name** permalink structures however I only posted a single video.

Here's a video of the table of contents working for me on initial save with the **Post Name** permalink setting selected.


https://user-images.githubusercontent.com/60436221/188423437-af756b90-8925-4efd-8362-49b337a4beba.mp4

**Comment by aatanasovdev** [comment]
Hi @aaronrobertshaw, it looks like you're adding the Table of Content block to a post, not a page. 

Publishing a post works slightly differently than a page. 

When we add a post, there is no auto-draft URL generation before publishing the page (I haven't checked why this happens, but it has to be consistent). In that case, the TOC block uses URLs like `SITE_URL/?p=34`. The links work because there is a redirect from `SITE_URL/?p=34` to the pretty URL (WordPress handles this redirect).

On the other side, when a page is published, the TOC block uses `auto-draft` inside its links.

Here is a video that shows how the block behaves.

[Test-TOC-block.webm](https://user-images.githubusercontent.com/38669058/188966364-b16c8d19-3080-443d-a6b5-6cf1bfceaf69.webm)

Probably, it will be better if the `auto-draft` URL generation issue is addressed. Initially, the links will use the plain format, which will be changed in the next content update. In that case, the links will work without pointing to a non-existing page.

Alternatively, the links can be regenerated automatically behind the scenes when a page is published. However, it doesn't seem right.

**Comment by aaronrobertshaw** [comment]
> it looks like you're adding the Table of Content block to a post, not a page.

🤦 Sorry to waste your time. You are correct, I was on autopilot and misread the instructions.

I can confirm I'm seeing this issue when I actually test it properly.

<img width="400" alt="Screen Shot 2022-09-08 at 10 26 08 am" src="https://user-images.githubusercontent.com/60436221/189007510-e6821485-2e7f-4c2f-b7e0-79c9fe3e6696.png">

**Comment by mahnunchik** [comment]
I can confirm incorrect behaviour:

* Incorrect URL in ToC block after initial save/publish action.
* After second `Update` all links are correct.

WordPress version `6.2.0`
Gutenberg plugin `15.1.1`

**Comment by felipevelzani** [comment]
I could reproduce this as well. Using Gutenberg 16.4.

This is the code on the page in case it helps: https://d.pr/n/kp52JN

**Comment by annezazu** [comment]
Noting that this was brought up in the [twenty fifth call for testing for the FSE Outreach Program using 16.4:](https://make.wordpress.org/test/2023/08/09/fse-program-testing-call-25-lets-start-from-the-beginning/#comment-3125)

>But it looks like the block does not work on previews only after the page has been published. As the preview does not have the correct structure. https://bluewinddesign.com/?page_id=385 (previewed page). Clicking one of the headings in the Table of Contents block the link anchor looks like this: https://bluewinddesign.com/resume/#head-3 . It went to a 404 page.

I've added it to the stabilizing the TOC issue!

**Comment by Sukhendu2002** [comment]
I attempted to reproduce this issue but was not able to—it appears to be resolved.

Tested with:
- WordPress: 6.9-alpha-60275
- Gutenberg: 21.0.0-rc.1

https://github.com/user-attachments/assets/88ac88cd-02c4-4453-ae33-c794ef5d793b

**Comment by t-hamano** [comment]
This problem has not yet been resolved. Two important things when testing are:

- Test with a new "Page" instead of a new "Post"
- Set the Permalink structure to "Post name"

**Comment by scruffian** [comment]
Fix here: https://github.com/WordPress/gutenberg/pull/71599

**Comment by Mamaduka** [comment]
I think instead of trying to resolve on the client side, we should append permalinks during render on the server and store relative URLs inside attributes. This solves a couple of problems:

1. Links should work for any post status.
2. ToC doesn't need to update when permalinks change.
3. Much easier to migrate the ToC block between sites.
4. Internal - we can do targeted subscription and only subscribe to the `block-editor` store.

We can extract and reuse rendering logic from my old PR - https://github.com/WordPress/gutenberg/pull/54224, or use HTML API and modify hrefs.

IMO, dynamic values like permalinks are always better to handle server-side; this is what we do in other blocks.

**Comment by scruffian** [comment]
Sorry closed accidentally

**Comment by mikachan** [comment]
I've attempted to fix this as part of https://github.com/WordPress/gutenberg/pull/72110.

**Comment by t-hamano** [comment]
Since the ToC block itself has been punted to WP 7.0, this issue will also be punted to 7.0.

**Comment by tyxla** [comment]
@getdave is this addressed after #80318 and #80404? If yes, let's close it, and also consider closing #70386.

**Comment by Mamaduka** [comment]
Yes, this was fixed when we switched the block to server-side rendering.
